README.md
# rethinkdb-ts-migrate
A migration tool for RethinkDB

## Overview

`rethinkdb-ts-migrate` is a utility allowing you to version
changes to a [RethinkDB](http://rethinkdb.com) database. Forward
or backward migrate structural or data changes. Share
and control these changes with other developers.

This code is a fork of Johan Obrink's
[rethink-migrate](https://github.com/JohanObrink/rethink-migrate).
`rethinkdb-ts-migrate` is more actively maintained.


## Install

```bash
$ npm install -g rethinkdb-ts-migrate
```

## Setup

Install a RethinkDB driver:

```bash
npm install --save rethinkdb
```

or

```bash
npm install --save rethinkdbdash
```

Create a ```database.json``` file in the root of your solution with the format:

```json
{
  "host": "localhost",
  "port": 28015,
  "db": "migrations",
  "discovery": true,
  "timeout": 60
}
```

Other, optional, parameters are `authKey` and `ssl`.

You can also use environment variables or arguments to override.

## Create migration

```bash
$ rethinkdb-ts-migrate create [migration name]
```

For example:

```bash
rethinkdb-ts-migrate create add-tables
```

## Edit migration

Open the file `./migrations/[timestamp]-[migration name].ts`

Add the changes to be made. For example:

```javascript
exports.up = function (r, connection) {
  return r.tableCreate('foo').run(connection);
};

exports.down = function (r, connection) {
  return r.tableDrop('foo').run(connection);
};
```

## Run migrations

`rethinkdb-ts-migrate up --one` will run one up migration.

`rethinkdb-ts-migrate up --all` will run all outstanding up migrations.

`rethinkdb-ts-migrate down --one` will run one down migration.

`rethinkdb-ts-migrate down --all` will run all outstanding down migrations.

If either `--all` or `--one` isn't specified, `--all` is the default.

### Options

`rethinkdb-ts-migrate up --root ./build` will run all outstanding up migrations
  found in `./build/migrations` with `database.json` in `./build`.
`-r` can be used as an alias.

`rethinkdb-ts-migrate up --logLevel debug` will set logLevel to debug.
  Possible values are: debug | info | warning | error | none.
`-l` can be used as an alias.

## Run tests

Run `docker-compose up`
If necessary, change the IP address in `test/database.json`
Run `gulp jshint test`

To run tests continually, add `watch` to the gulp command

# License

The MIT License (MIT)

Copyright (c) 2017 [Parabol, Inc.](https://parabol.co)
Copyright (c) 2015 Johan Öbrink

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

This CHANGELOG follows conventions [outlined here](http://keepachangelog.com/).

## 0.3.5 - 04-Nov-2022

- Bump nconf to not use dynamic requires

## 0.3.2 - 04-Jan-2022

### Fixed

- Security patches

## 0.3.1 - 04-Jan-2022

### Fixed

- Security patches

## v0.1.0 - 170-Oct-2019

### Added

- Support for typescript migration files


src/migrate.ts
import chalk from 'chalk'
import {promises as fs} from 'fs'
import moment from 'moment'
import nconf from 'nconf'
import * as path from 'path'
import {r, R} from 'rethinkdb-ts'

require('sucrase/register')

const MIGRATION_TABLE_NAME = '_migrations'
const rxMigrationFile = /^\d{14}-.*\.(js|ts)$/
type MigrationFn = (r: R) => void

interface MigrationFile {
  up: MigrationFn
  down: MigrationFn
}

/*
  Read config from
  - arguments
  - environment variables
  - /database.json
 */
type Unpromise<T> = T extends (...args: any[]) => Promise<infer U> ? U : T

const getConfig = async (root: string) => {
  nconf
    .argv()
    .env()
    .file({file: path.join(root, 'database.json')})
  return {
    host: nconf.get('host'),
    port: nconf.get('port'),
    user: nconf.get('user') || 'admin',
    password: nconf.get('password') || '',
    db: nconf.get('db'),
    discovery: Boolean(nconf.get('discovery')) || false,
    timeout: nconf.get('timeout') || 5 * 60,
    authKey: nconf.get('authKey'),
    ssl: nconf.get('ssl'),
  }
}
type Config = Unpromise<typeof getConfig>

/*
  Connect to db
  If db does not yet exist, create it
 */
async function connectToDb(config: Config) {
  await r.connectPool(config)
  const dbs = await r.dbList().run()
  if (!dbs.includes(config.db)) {
    await r.dbCreate(config.db).run()
  }
  const tables = await r.tableList().run()
  if (!tables.includes(MIGRATION_TABLE_NAME)) {
    await r.tableCreate(MIGRATION_TABLE_NAME).run()
    await r.table(MIGRATION_TABLE_NAME).indexCreate('timestamp').run()
    await r.table(MIGRATION_TABLE_NAME).indexWait().run()
  }
  await r.db(config.db).wait({waitFor: 'ready_for_writes'}).run()
}

/*
  Compares completed migrations to files on disk
  Returns the migrations scripts with a timestamp newer than last
  completed migration in db.

  TODO: Change so that all non run migration scripts are returned
 */
async function getMigrationsExcept(
  completedMigration: Migration | undefined,
  root: string,
  numToApply: -1 | 1,
) {
  const dir = path.join(root, 'migrations')
  const files = await fs.readdir(dir)
  const migrationFiles = files.filter((file) => file.match(rxMigrationFile))
  const migrations = migrationFiles
    .map((filename) => {
      const tsix = filename.indexOf('-')
      return {
        name: filename.substring(tsix + 1, filename.lastIndexOf('.')),
        timestamp: filename.substring(0, tsix),
        filename: filename,
      }
    })
    .filter((migration) => {
      return completedMigration ? migration.timestamp > completedMigration.timestamp : true
    })
  const migrationsToApply = numToApply !== -1 ? migrations.slice(0, numToApply) : migrations
  const codes = await Promise.all(
    migrationsToApply.map(
      (migration) => require(path.join(dir, migration.filename)) as MigrationFile,
    ),
  )
  return migrationsToApply.map((migration, idx) => ({
    ...migration,
    code: codes[idx],
  }))
}

/*
  Takes a list of migration file paths and requires them
 */
function requireMigrations(migrations: Migration[], root: string) {
  return migrations.map(function (migration) {
    const filename = migration.timestamp + '-' + migration.name
    const filepath = path.join(root, 'migrations', filename)
    return {
      ...migration,
      filename,
      code: require(filepath) as MigrationFile,
    }
  })
}

/*
  Run all new up migrations
 */

interface Params {
  root?: string
  all?: boolean
}

interface Migration {
  id: string
  name: string
  timestamp: string
}

export async function up(params: Params): Promise<void> {
  const {all, root} = params
  const rootDir = root || process.cwd()
  await connectToDb(await getConfig(rootDir))
  const completedMigrations = await r
    .table<Migration>(MIGRATION_TABLE_NAME)
    .orderBy({index: 'timestamp'})
    .run()
  const latest = completedMigrations[completedMigrations.length - 1]
  const numToApply = all ? -1 : 1
  const migrations = await getMigrationsExcept(latest, rootDir, numToApply)
  if (migrations.length < 1) {
    logInfo('No new migrations')
    await r.getPoolMaster()?.drain()
    return
  }
  for (let i = 0; i < migrations.length; i++) {
    const migration = migrations[i]
    logInfo(chalk.black.bgGreen(' ↑  up  ↑ '), migration.timestamp, chalk.yellow(migration.name))
    const {name, timestamp} = migration
    try {
      await migration.code.up(r)
    } catch (e) {
      logError(`Migration failed: ${name}`, e as any)
      await r.getPoolMaster()?.drain()
      return
    }
    await r.table(MIGRATION_TABLE_NAME).insert({name, timestamp}).run()
  }
  logInfo('Migration Successful')
  await r.getPoolMaster()?.drain()
}

/*
  Rollback one or all migrations
 */
export async function down(params: Params): Promise<void> {
  const {all, root} = params
  const rootDir = root || process.cwd()
  await connectToDb(await getConfig(rootDir))
  const completedMigrations = await r
    .table<Migration>(MIGRATION_TABLE_NAME)
    .orderBy({index: 'timestamp'})
    .run()
  if (completedMigrations.length === 0) {
    logInfo('No new migrations')
    await r.getPoolMaster()?.drain()
    return
  }
  const migrationsToRollBack = all
    ? completedMigrations.reverse()
    : [completedMigrations[completedMigrations.length - 1]]
  const migrations = requireMigrations(migrationsToRollBack, rootDir)
  for (let i = 0; i < migrations.length; i++) {
    const migration = migrations[i]
    logInfo(chalk.black.bgYellow(' ↓ down ↓ '), migration.timestamp, chalk.yellow(migration.name))
    const {name} = migration
    try {
      await migration.code.down(r)
    } catch (e) {
      logError(`Migration failed: ${name}`, e as any)
      await r.getPoolMaster()?.drain()
      return
    }
    await r.table(MIGRATION_TABLE_NAME).get(migration.id).delete().run()
  }
  logInfo('Migration successful')
  await r.getPoolMaster()?.drain()
}

function logInfo(...args: any[]) {
  args.unshift(chalk.blue('[migrate-rethinkdb]'))
  console.log(args.join(' '))
}

function logError(txt: string, error: Error) {
  console.error(chalk.red('[migrate-rethinkdb] ') + txt)
  if (error) {
    console.error(error)
    if (error.stack) {
      console.error(error.stack)
    }
  }
}

export async function create(name: string, root: string): Promise<string> {
  const rootDir = path.join(root || process.cwd(), 'migrations')
  const templatepath = path.join(__dirname, '../template.ts')
  const filename = moment().format('YYYYMMDDHHmmss') + '-' + name + '.ts'
  const filepath = path.join(rootDir, filename)
  await fs.copyFile(templatepath, filepath)
  return filepath
}


lib/migrate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.create = exports.down = exports.up = void 0;
const tslib_1 = require("tslib");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const fs_1 = require("fs");
const moment_1 = tslib_1.__importDefault(require("moment"));
const nconf_1 = tslib_1.__importDefault(require("nconf"));
const path = tslib_1.__importStar(require("path"));
const rethinkdb_ts_1 = require("rethinkdb-ts");
require('sucrase/register');
const MIGRATION_TABLE_NAME = '_migrations';
const rxMigrationFile = /^\d{14}-.*\.(js|ts)$/;
const getConfig = async (root) => {
    nconf_1.default
        .argv()
        .env()
        .file({ file: path.join(root, 'database.json') });
    return {
        host: nconf_1.default.get('host'),
        port: nconf_1.default.get('port'),
        user: nconf_1.default.get('user') || 'admin',
        password: nconf_1.default.get('password') || '',
        db: nconf_1.default.get('db'),
        discovery: Boolean(nconf_1.default.get('discovery')) || false,
        timeout: nconf_1.default.get('timeout') || 5 * 60,
        authKey: nconf_1.default.get('authKey'),
        ssl: nconf_1.default.get('ssl'),
    };
};
/*
  Connect to db
  If db does not yet exist, create it
 */
async function connectToDb(config) {
    await rethinkdb_ts_1.r.connectPool(config);
    const dbs = await rethinkdb_ts_1.r.dbList().run();
    if (!dbs.includes(config.db)) {
        await rethinkdb_ts_1.r.dbCreate(config.db).run();
    }
    const tables = await rethinkdb_ts_1.r.tableList().run();
    if (!tables.includes(MIGRATION_TABLE_NAME)) {
        await rethinkdb_ts_1.r.tableCreate(MIGRATION_TABLE_NAME).run();
        await rethinkdb_ts_1.r.table(MIGRATION_TABLE_NAME).indexCreate('timestamp').run();
        await rethinkdb_ts_1.r.table(MIGRATION_TABLE_NAME).indexWait().run();
    }
    await rethinkdb_ts_1.r.db(config.db).wait({ waitFor: 'ready_for_writes' }).run();
}
/*
  Compares completed migrations to files on disk
  Returns the migrations scripts with a timestamp newer than last
  completed migration in db.

  TODO: Change so that all non run migration scripts are returned
 */
async function getMigrationsExcept(completedMigration, root, numToApply) {
    const dir = path.join(root, 'migrations');
    const files = await fs_1.promises.readdir(dir);
    const migrationFiles = files.filter((file) => file.match(rxMigrationFile));
    const migrations = migrationFiles
        .map((filename) => {
        const tsix = filename.indexOf('-');
        return {
            name: filename.substring(tsix + 1, filename.lastIndexOf('.')),
            timestamp: filename.substring(0, tsix),
            filename: filename,
        };
    })
        .filter((migration) => {
        return completedMigration ? migration.timestamp > completedMigration.timestamp : true;
    });
    const migrationsToApply = numToApply !== -1 ? migrations.slice(0, numToApply) : migrations;
    const codes = await Promise.all(migrationsToApply.map((migration) => require(path.join(dir, migration.filename))));
    return migrationsToApply.map((migration, idx) => ({
        ...migration,
        code: codes[idx],
    }));
}
/*
  Takes a list of migration file paths and requires them
 */
function requireMigrations(migrations, root) {
    return migrations.map(function (migration) {
        const filename = migration.timestamp + '-' + migration.name;
        const filepath = path.join(root, 'migrations', filename);
        return {
            ...migration,
            filename,
            code: require(filepath),
        };
    });
}
async function up(params) {
    const { all, root } = params;
    const rootDir = root || process.cwd();
    await connectToDb(await getConfig(rootDir));
    const completedMigrations = await rethinkdb_ts_1.r
        .table(MIGRATION_TABLE_NAME)
        .orderBy({ index: 'timestamp' })
        .run();
    const latest = completedMigrations[completedMigrations.length - 1];
    const numToApply = all ? -1 : 1;
    const migrations = await getMigrationsExcept(latest, rootDir, numToApply);
    if (migrations.length < 1) {
        logInfo('No new migrations');
        await rethinkdb_ts_1.r.getPoolMaster()?.drain();
        return;
    }
    for (let i = 0; i < migrations.length; i++) {
        const migration = migrations[i];
        logInfo(chalk_1.default.black.bgGreen(' ↑  up  ↑ '), migration.timestamp, chalk_1.default.yellow(migration.name));
        const { name, timestamp } = migration;
        try {
            await migration.code.up(rethinkdb_ts_1.r);
        }
        catch (e) {
            logError(`Migration failed: ${name}`, e);
            await rethinkdb_ts_1.r.getPoolMaster()?.drain();
            return;
        }
        await rethinkdb_ts_1.r.table(MIGRATION_TABLE_NAME).insert({ name, timestamp }).run();
    }
    logInfo('Migration Successful');
    await rethinkdb_ts_1.r.getPoolMaster()?.drain();
}
exports.up = up;
/*
  Rollback one or all migrations
 */
async function down(params) {
    const { all, root } = params;
    const rootDir = root || process.cwd();
    await connectToDb(await getConfig(rootDir));
    const completedMigrations = await rethinkdb_ts_1.r
        .table(MIGRATION_TABLE_NAME)
        .orderBy({ index: 'timestamp' })
        .run();
    if (completedMigrations.length === 0) {
        logInfo('No new migrations');
        await rethinkdb_ts_1.r.getPoolMaster()?.drain();
        return;
    }
    const migrationsToRollBack = all
        ? completedMigrations.reverse()
        : [completedMigrations[completedMigrations.length - 1]];
    const migrations = requireMigrations(migrationsToRollBack, rootDir);
    for (let i = 0; i < migrations.length; i++) {
        const migration = migrations[i];
        logInfo(chalk_1.default.black.bgYellow(' ↓ down ↓ '), migration.timestamp, chalk_1.default.yellow(migration.name));
        const { name } = migration;
        try {
            await migration.code.down(rethinkdb_ts_1.r);
        }
        catch (e) {
            logError(`Migration failed: ${name}`, e);
            await rethinkdb_ts_1.r.getPoolMaster()?.drain();
            return;
        }
        await rethinkdb_ts_1.r.table(MIGRATION_TABLE_NAME).get(migration.id).delete().run();
    }
    logInfo('Migration successful');
    await rethinkdb_ts_1.r.getPoolMaster()?.drain();
}
exports.down = down;
function logInfo(...args) {
    args.unshift(chalk_1.default.blue('[migrate-rethinkdb]'));
    console.log(args.join(' '));
}
function logError(txt, error) {
    console.error(chalk_1.default.red('[migrate-rethinkdb] ') + txt);
    if (error) {
        console.error(error);
        if (error.stack) {
            console.error(error.stack);
        }
    }
}
async function create(name, root) {
    const rootDir = path.join(root || process.cwd(), 'migrations');
    const templatepath = path.join(__dirname, '../template.ts');
    const filename = (0, moment_1.default)().format('YYYYMMDDHHmmss') + '-' + name + '.ts';
    const filepath = path.join(rootDir, filename);
    await fs_1.promises.copyFile(templatepath, filepath);
    return filepath;
}
exports.create = create;
//# sourceMappingURL=migrate.js.map